### PR TITLE
Gale Harrington - Carets

### PR DESCRIPTION
**Pull request by galestorm**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | One decision I made was to have a method in my Block class that just transfers a Room object from a list of rooms in the block to a list of "off-limits" (already reserved) rooms in that block. I had considered making changes to my Room class that would have changed the criteria for availability or would have given Room some knowledge of whether it was in a block. In the end, I just made the change that to me seemed easiest and most clear (writing new code rather than going back to adjust my Room code).  |
| Describe a concept that you gained more clarity on as you worked on this assignment. | I got more clarity on how different classes need to relate to one another in order for them to "communicate". I got more comfortable figuring out what information classes need to contain themselves vs. what information they just need to access.   |
| Describe a _nominal test_ that you wrote for this assignment. | For hotel_spec.rb, I made sure that if I reserved a single room for a specific date range, my array of reservations increased by one, as I would expect.  |
| Describe an _edge case test_ that you wrote for this assignment. | Also in hotel_spec.rb, I wrote a test to ensure that attempting to create a block when there were no available rooms would raise an error.    |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I didn't really write a lot of pseudo-code, but I was really happy I took the time to think about the design of each wave before I touched code. I think that saved me from a lot of frustration. I also tried to do the bare minimum for each user story, and didn't look ahead at all, which kept me from overcomplicating the assignment. I continue to think TDD is the bomb for tackling one little problem at a time, but I suspect I might not be doing a great job of testing edge cases because I get too excited when my nominal tests pass.  |

### CODE REVIEW

**Comment by tildeee** [comment]
Good code! I think calculating reservations from the stored array of rooms is very interesting and works out very well.

By the way, I LOVE the `lets` that you used in the tests!

In terms of (Wave 3) requirements, I would've loved to see the Block functionality such that from hotel I could reserve a number of rooms that were in a block.

File: lib/reservation.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+
+module Hotel_System
+  class Reservation
+    attr_accessor :check_in, :check_out, :room, :rate_adjustor
+    def initialize(check_in, check_out)
+      raise ArgumentError.new("Check_in Date object, check_out Date object") unless check_in.class == Date && check_out.class == Date
+      raise ArgumentError.new("Cannot check in and out on same day") if check_in == check_out
+      raise ArgumentError.new("Time Warp") if (check_out <=> check_in) == -1
```
Comment (tildeee) [inline lib/reservation.rb:8]: There's some duplication for the Date logic/error handling between Reservation and Room. I wish there were a way you could pull this out to something reused, or at least placed in the same place!

File: lib/room.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      @room_number = room_number
+      @reservations = []
+      @rate_adjustor = rate_adjustor
+    end
+
+    def available?(check_in, check_out)
+      overlap = []
+      (check_in...check_out).each do |date|
+        overlap += self.reservations.select do |reservation|
+          (reservation.check_in...reservation.check_out).cover?(date)
```
Comment (tildeee) [inline lib/room.rb]: Not a big deal at all because I think it's fairly contained in this project, BUT, there could be some refactoring with this line.

Hotel is reaching into each reservation and performing the logic on if the reservation is in this date. It might be interesting to pull this into a small helper function so that reservation does it itself, so that in hotel it could be:
```
reservation.covers?(date)
```
and in `Reservation` there's a method
```
def covers? date
  (@check_in...@check_out).cover?(date)
end
```

File: lib/hotel.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+          reservations += room.reservations
+        end
+      end
+      return reservations
+    end
+
+    def find_res_by_date(date)
+      all_reservations = self.reservations
+      date_reservations = all_reservations.select do |reservation|
+        (reservation.check_in...reservation.check_out).cover?(date)
```
Comment (tildeee) [inline lib/hotel.rb]: Not a big deal at all because I think it's fairly contained in this project, BUT, there could be some refactoring with this line.

Hotel is reaching into each reservation and performing the logic on if the reservation is in this date. It might be interesting to pull this into a small helper function so that reservation does it itself, so that in hotel it could be:
```
reservation.covers?(date)
```
and in `Reservation` there's a method
```
def covers? date
  (@check_in...@check_out).cover?(date)
end
```

**Comment by tildeee** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| x
Answer comprehension questions	| x
**Design** |
Each class is responsible for a single piece of the program | x
Classes are loosely coupled | x
**Wave 1** |
List rooms | x
Reserve a room for a given date range | x
List reservations for a given date | x
Calculate reservation price | x
Invalid date range produces an error | see comments about refactoring
Test coverage | x
**Wave 2** |
View available rooms for a given date range | x
Reserving a room that is not available produces an error | x
Test coverage | x
**Wave 3** |
Create a block of rooms | x
Check if a block has rooms | x
Reserve a room from a block | ?
Test coverage | x
**Additional Feedback** |
